library/alloc/src/vec/in_place_collect.rs

```diff
@@ -186,6 +186,32 @@ const fn in_place_collectible<DEST, SRC>(
     }
 }
 
+const fn needs_realloc<SRC, DEST>(src_cap: usize, dst_cap: usize) -> bool {
+    if const { mem::align_of::<SRC>() != mem::align_of::<DEST>() } {
+        return true;
+    }
+
+    if const {
+        // examples that may require reallocs unless src/dest capacities turn out to be multiples at runtime
+        // Vec<u8> -> Vec<[u8; 4]>
+        let dst_larger = mem::size_of::<SRC>() < mem::size_of::<DEST>();
+
+        // Vec<[u8; 3]> -> Vec<[u8; 2]>
+        // dest_sz can't actually be 0 since in_place_collectible() checks for that but const eval doesn't know that
+        let dest_sz = mem::size_of::<DEST>();
+        let src_not_multiple_of_dest =
+            dest_sz == 0 || mem::size_of::<SRC>() % mem::size_of::<DEST>() != 0;
```

Comment:
Why does this have the `dest_sz == 0` check at all if it's not possible to be 0? IMO it would be more desirable for const eval to panic in case the mentioned check gets removed by mistake (especially since it's not local) and would be clearer to not have that code path if it's meant to be a constant `false || ...`.

Reply:
because the code path still gets instantiated. it just becomes unreachable due to an earlier branch on `in_place_collectible()`, which means the consts still get evaluated. If I panic here it panics at compile time even though the branch is dead.